cmd/dptp-controller-manager/main.go

```diff
@@ -236,24 +234,21 @@ func main() {
 	ctx := controllerruntime.SetupSignalHandler()
 	ctx, cancel := context.WithCancel(ctx)
 
-	kubeconfigChangedCallBack := func(e fsnotify.Event) {
-		logrus.WithField("event", e.String()).Info("Kubeconfig changed, exiting to get restarted by Kubelet and pick up the changes")
+	kubeconfigChangedCallBack := func() {
+		logrus.Info("Kubeconfig changed, exiting to get restarted by Kubelet and pick up the changes")
 		cancel()
 	}
 
-	kubeconfigs, err := util.LoadKubeConfigs(opts.kubeconfig, opts.kubeconfigDir, kubeconfigChangedCallBack)
+	kubeconfigs, err := opts.kubernetesOptions.LoadClusterConfigs(kubeconfigChangedCallBack)
 	if err != nil {
 		logrus.WithError(err).Fatal("failed to load kubeconfigs")
 	}
 	if _, hasAppCi := kubeconfigs[appCIContextName]; !hasAppCi {
```

Comment:
Do we need this branch? What do we do in the deployment? I don't understand why our loading has to be so complex. Either always expect it to be in-cluster, or always expect it to be present in the configs ...

Reply:
It turns out we need both:

development: load from $KUBECONFIG
https://github.com/openshift/ci-tools/blob/12fad0a042c90eda3cddf44fb8dd84a1ea6ea9e6/hack/run-dptp-controller-manager.sh#L11

production: load from inClusterConfig because the secret does not include the one for `app.ci`:

https://github.com/openshift/release/blob/fa66cd849fb407495cd50fab336d3f0e88fcb622/core-services/ci-secret-bootstrap/_config.yaml#L1948-L1961

Maybe add it into the secret and then remove the code for inClusterConfig?

Reply:
Change development. We do not write code to handle development situations, we write it to handle production. Developing on different code-paths to production is a recipe for disaster.

Reply:
The k8s loader also allows you to provide it a loading strategy, so that it perfers one to the other. @alvaroaleman is there any simpler way to do this? I really dislike how damn complicated we've made loading. The library is complex as all get-out, then each caller does stuff like this to make it even more impossible.

Reply:
This uses the in-cluster config for app.ci if there is no app.ci config. I don't really see how anything in client-go would be able to do this, as it doesn't know what name the in-cluster config is for.

I also don't really see how any change in the deployment would improve things here. Sure, we could add a flag for `--default-in-cluster-config-for-cluster-name` but I don't anticipate this changing so it looks more like busywork than like something that improves things to me.

Reply:
I just hate that we have 35 places that load config, and their call-sites are all over the place. Half of them do something like this, half don't, half have custom handling for different things, half don't ... next time anyone has to do a broad change it's going to be as much of a nightmare as it has been for Hongkai. Surely we can do better?

Reply:
Maybe load in-cluster if possible, add it to a well-known key in the output mapping that the upstream library provides, then the caller just moves map keys around?

Reply:
Updated.
Not sure if it is getting better or worse though.

Reply:
Do we put the in-cluster config into a known key? Can we fail over to that? @alvaroaleman any thoughts?

Reply:
Yeah we do put it into a known key, it is handled above